EXILED/Exiled.API/Features/Core/UserSettings/SettingGroup.cs

```diff
@@ -0,0 +1,159 @@
+        public static implicit operator SettingGroup(SettingBase[] settings) => new(settings);
+
+        /// <summary>
+        /// Recursively gets all instances of <see cref="SettingBase"/> contained within this <see cref="SettingGroup"/>'s Settings and SubGroups.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerable{T}"/> of <see cref="SettingBase"/>.</returns>
+        public IEnumerable<SettingBase> GetAllSettings()
+        {
+            List<SettingBase> settings = new();
+            List<SettingGroup> previousGroups = ListPool<SettingGroup>.Pool.Get(1);
+            InternalGetAllSettings(settings, previousGroups);
+            ListPool<SettingGroup>.Pool.Return(previousGroups);
+            return settings;
+        }
+
+        /// <summary>
+        /// Recursively gets all instances of <see cref="SettingBase"/> contained within this <see cref="SettingGroup"/>'s Settings and SubGroups that can be viewed by a specified player.
+        /// </summary>
+        /// <param name="viewer"><inheritdoc cref="Player"/></param>
+        /// <returns>An <see cref="IEnumerable{T}"/> of <see cref="SettingBase"/>.</returns>
+        public IEnumerable<SettingBase> GetViewableSettingsOrdered(Player viewer)
+        {
+            if (viewer == null)
+                return Array.Empty<SettingBase>();
+            List<SettingBase> settings = new();
+            List<SettingGroup> previousGroups = ListPool<SettingGroup>.Pool.Get(1);
+            InternalGetViewableSettingsOrdered(settings, previousGroups, viewer);
+            ListPool<SettingGroup>.Pool.Return(previousGroups);
+            return settings;
+        }
+
+        /// <summary>
+        /// Returns a string representation of this <see cref="SettingGroup"/>.
+        /// </summary>
+        /// <returns>A string in human-readable format.</returns>
+        public override string ToString() => $"{Priority} ({Viewers}) [{string.Join(", ", Settings.Select(s => s.ToString()))}]";
+
+        /// <summary>
+        /// An internal recursive method to get all settings within a SettingGroup.
+        /// </summary>
+        /// <param name="current">A <see cref="List{T}"/> of <see cref="SettingBase"/> to add new settings to.</param>
+        /// <param name="previousGroups">A <see cref="List{T}"/> of <see cref="SettingGroup"/> used to verify no recursive loops.</param>
+        internal void InternalGetAllSettings(List<SettingBase> current, List<SettingGroup> previousGroups)
+        {
+            current.AddRange(Settings);
+            if (SubGroups == null)
+                return;
+            previousGroups.Add(this);
+            foreach (SettingGroup group in SubGroups)
+            {
+                if (previousGroups.Contains(group))
+                    throw new InvalidOperationException("SettingGroups cannot reference themselves within subgroups.");
+                group.InternalGetAllSettings(current, previousGroups);
+            }
+        }
+
```

Comment:
idk if I'm going schizo with recursive thoughts n stuff, but wouldn't we want to remove the first instance of the current group after we foreach in all the sub-groups? That way you could reference a SettingGroup several times in SubGroups within multiple other groups? Just a consideration. Rn this would work with like 99% of applications. I'm just trying to make that other 1% possible before this gets merged or anything
```suggestion
        internal void InternalGetAllSettings(List<SettingBase> current, List<SettingGroup> previousGroups)
        {
            current.AddRange(Settings);
            if (SubGroups == null)
                return;
            previousGroups.Add(this);
            foreach (SettingGroup group in SubGroups)
            {
                if (previousGroups.Contains(group))
                    throw new InvalidOperationException("SettingGroups cannot reference themselves within subgroups.");
                group.InternalGetAllSettings(current, previousGroups);
            }
            previousGroups.Remove(this);
        }
```
Would this be better while still checking for recursive loops?

Reply:
also idk how to comment on multiple lines like VALERA, I'm too stupid 😔